### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by LuKuangChen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +90,44 @@ found 1725 well-typed programs.
...
+```
+
+> Static Python has 537 other tests (802 total) that we did not use because they fall outside the scope of the model.
+
+`make statistics` also produces this number (802).
+
+> Most [casts] call for a tag check, i.e., a Python isinstance test. The sole exception is optional types, which require a tag check and a test for the none value.
+
+If you look at [the definition of `compile-check` in `./compile.rkt`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/compile.rkt#L1225), the `dynamic` case is trivial. The function case is unreachable because there is no function type in the surface syntax. We use function types internally to make checking function/method calls easier. The remaining two cases are more interesting. When the target type is a class, `compile-check` calls [`check-exactness`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/compile.rkt#L1244), which effectively performs a tag-check. The remaining case is for `Optional[T]`, which has a test for the none value and a recursive call. Because the `T` in `Optional[T]` must be a class, the check must be a tag check.
```
[the remark this answers is not in the document]
Comment (bennn) [reply README.md:125]: > a test for the none value and a recursive call

Replace "recursive call" with "tag check" or "recursive tag check".

With the current two sentences, it looks like the overall check for `Optional[T]` is a tag check.

(Another fix would be to change the final sentence: replace "the check must be" with "the recursive call must perform". But I'd rather we keep the text as short as possible.)

**Review by bennn — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -90,6 +90,44 @@ found 1725 well-typed programs.
 </pre>
 </details>
 
+## How to validate claims from the paper?
+
```
[the remark this answers is not in the document]
Comment (LuKC1024) [reply README.md:94]: Fixed by [f07d5df](https://github.com/brownplt/insta-model/pull/3/commits/f07d5df017c79eef604335740bda39697e6c5ef3)

File: README.md [hunk 3]
```diff
@@ -90,6 +90,44 @@ found 1725 well-typed programs.
...
+> The model covers a substantial part of the Python language including assertions, loops, exception handlers, and delete statements.
+
+It is easy to find those language constructs in [./conformance_suite](./conformance_suite).
```
[the remark this answers is not in the document]
Comment (LuKC1024) [reply README.md]: That makes a lot of sense. Fixed in 0a5a73e and 67eb471.

File: README.md [hunk 4]
```diff
@@ -90,6 +90,44 @@ found 1725 well-typed programs.
...
+
+> Second, we used Redex’s random testing tools [kf-sfp-2009] to check type soundness on thousands of examples (1,600 expressions and 11,000 programs).
+
+We have two uses of [`redex-check`](https://docs.racket-lang.org/redex/reference.html#%28form._%28%28lib._redex%2Freduction-semantics..rkt%29._redex-check%29%29) in [./conjectures.rkt](./conjectures.rkt). (TODO: numbers in code are different from numbers in paper. What is the right thing to do?)
```
[the remark this answers is not in the document]
Comment (LuKC1024) [reply README.md]: Resolved.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (LuKC1024) [reply README.md]: Fixed in d071cbf890ad722b2508c0a0ed9f1c5cb307ec29

File: README.md [hunk 5]
```diff
@@ -90,6 +90,63 @@ found 1725 well-typed programs.
...
+We will walk you through the first use. The other is very similar. On line 172 you will see the following expression. The expression generate 10000 random expressions (`e+`) and checks whether for all expression `e+`, if the expression compiles and terminates within certain amoount of reduction steps, the result must be of the same type as the expression. The nonterminate `e+` is defined in language `SP-conjecture`, which inherite the definition from language `SP` ([line 29 of `grammar.rkt`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/grammar.rkt#L29)
+
+```racket
+(redex-check SP-conjecture
+             e+
+             (term (compile-implies-terminate-implies-well-typed-e e+))
+             #:attempts 10000)
+```
+
+In the repository we tested 10000 random expressions and 15000 random programs. The numbers are different in the paper (1600 and 11000) but you can always change the numbers.
```
Comment (bennn) [inline README.md:120]: The numbers in the paper are about **well-typed** terms, not random terms.
(I'll reword this).

File: README.md [hunk 6]
```diff
@@ -90,6 +90,63 @@ found 1725 well-typed programs.
...
+
+`make statistics` also produces this number (802).
+
+> (page 13; Section 4.2.1) Most [casts] call for a tag check, i.e., a Python isinstance test. The sole exception is optional types, which require a tag check and a test for the none value.
+
+If you look at [the definition of `compile-check` in `./compile.rkt`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/compile.rkt#L1225), the `dynamic` case is trivial. The function case is unreachable because there is no function type in the surface syntax. That's what SP does! (We introduced function types internally to make checking function/method calls easier.) The remaining two cases are more interesting. When the target type is a class, `compile-check` calls [`check-exactness`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/compile.rkt#L1244), which effectively performs a tag-check. The remaining case is for `Optional[T]`, which has a test for the none value and a tag check. Because the `T` in `Optional[T]` must be a class, the check must be a tag check.
```
Comment (bennn) [inline README.md:144]: "We introduced ...." makes it sound like SP doesn't have function types internally. But it does!
(I'll fix)

**Review by bennn — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -90,6 +90,63 @@ found 1725 well-typed programs.
...
+
+We have two uses of [`redex-check`](https://docs.racket-lang.org/redex/reference.html#%28form._%28%28lib._redex%2Freduction-semantics..rkt%29._redex-check%29%29) in [./conjectures.rkt](./conjectures.rkt) (line [172](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/conjectures.rkt#L172) and [239](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/conjectures.rkt#L239)).
+
+We will walk you through the first use. The other is very similar. On line 172 you will see the following expression. The expression generate 10000 random expressions (`e+`) and checks whether for all expression `e+`, if the expression compiles and terminates within certain amoount of reduction steps, the result must be of the same type as the expression. The nonterminate `e+` is defined in language `SP-conjecture`, which inherite the definition from language `SP` ([line 29 of `grammar.rkt`](https://github.com/brownplt/insta-model/blob/c9f21f5479b2dd4f9ddbfabacba88d22b3cb1811/grammar.rkt#L29)
```
Comment (bennn) [inline README.md:111]: heads up: "nonterminate `e+`" should be "nonterminal `e+`"
(I'll fix)

**Pull request merged**